### DISCUSSION TITLE
Runtime error with ssr noExternal differs in production from dev

### DISCUSSION THREAD

**Discussion by schaschko** [vikejs/vike#901] [Discussion] [answered]
Reproduce error:

`yarn create vite-plugin-ssr` -> react-ts
`cd vite-ssr-project/`
`yarn add @mui/material @emotion/react @emotion/styled @mui/icons-material`

Modify `pages/index/index.page.tsx`:

```
import { Counter } from './Counter'
import MenuIcon from '@mui/icons-material/Menu'; // add this

export { Page }

function Page() {
  return (
    <>
      <MenuIcon /> // and this
      <h1>Welcome</h1>
      This page is:
      <ul>
        <li>Rendered to HTML.</li>
        <li>
          Interactive. <Counter />
        </li>
      </ul>
    </>
  )
}
```

`yarn dev` works, `yarn prod` will throw runtime error when navigating to `localhost:3000`:

`Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.`

This error can be "fixed" by adding

```
ssr: {noExternal: ["@mui/material", "@mui/utils", "@mui/base", "@mui/icons-material"]},
```

to `vite.config.ts`. Unfortunately now `yarn dev` is broken:

```
7:13:10 PM [vite] Error when evaluating SSR module /node_modules/@mui/icons-material/Menu.js:
|- TypeError: require is not a function
```

What is going on here? It should work in production the same as in dev?

It is possible to do a different "fix":

In `pages/index/index.page.tsx` change `import MenuIcon from '@mui/icons-material/Menu';` to `import { Menu as MenuIcon } from '@mui/icons-material';` In this case, even `ssr: {noExternal: ["@mui/material", "@mui/utils", "@mui/base", "@mui/icons-material"]},` is not needed. It works in `dev` and `prod`. Problem is, my machine is running out of memory then, so this fix is not possible.
 
How to fix this best?

**Comment by brillout** [comment]
There isn't much VPS can do here. Can you report this to the `@mui` team? It seems like they wrongly published their pacakge.

**Reply by schaschko** [reply]
Thanks for looking into it, [I did](https://github.com/mui/material-ui/issues/37375).

**Reply by brillout** [reply]
I did some digging. While VPS can workaround the `require` issue, it doesn't solve the problem that the file `/node_modules/.pnpm/@mui+icons-material@5.11.16_@mui+material@5.13.2_@types+react@18.2.6_react@18.2.0/node_modules/@mui/icons-material/Menu.js` is CJS (Vite expects it to be ESM).

It seems like only the MUI team can solve the problem.

**Reply by brillout** [reply, marked as answer]
User land workaround: https://github.com/brillout/vps-mui/commit/89dd9925276ad8dab2238761332fb5f8e5efb407.

**Reply by brillout** [reply]
(Btw. in case you can convince your employer: https://github.com/sponsors/brillout.)

**Reply by michaldudak** [reply]
Hi, MUI maintainer here. For historical reasons (https://github.com/mui/material-ui/pull/26310) we don't ship ESM by default. We do, however, ship an ESM-ish bundle under the "esm" directory (@mui/icons-material/esm). This is the same format used by default in our other packages. A resolver alias could help here (see https://mui.com/material-ui/guides/minimizing-bundle-size/#available-bundles).

We plan to investigate shipping standards-compliant ESM for all the packages in the next major release.

**Reply by brillout** [reply]
@michaldudak Curious: is there a reason you don't ship ESM files only but as `.mjs` instead of `type: "module"`? That way it isn't a breaking change. I'm considering writing a guide for how to publish valid non-breaking ESM code.

**Reply by michaldudak** [reply]
We still [support Node 12](https://mui.com/material-ui/getting-started/supported-platforms/#server), so we can't ship only ES modules just yet.

**Reply by Maikpwwq** [reply]
Still facing same issue. I am currently getting this error:

``` 
...
[vite-plugin-vercel] Directory import '/media/oem/MyFiles/8_DEVELOPMENT/Comunidad-Dezzpo-PWA/node_modules/react-bootstrap/Row' is not supported resolving ES modules imported from /media/oem/MyFiles/8_DEVELOPMENT/Comunidad-Dezzpo-PWA/dist/server/entries/src_pages_app_biblioteca_index-page.mjs
Did you mean to import react-bootstrap@2.8.0_@types+react@18.2.18_react-dom@18.2.0_react@18.2.0/node_modules/react-bootstrap/cjs/Row.js?
✓ built in 1.34s
(node:31656) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/media/oem/MyFiles/8_DEVELOPMENT/Comunidad-Dezzpo-PWA/node_modules/.pnpm/@mui+material@5.14.3_@emotion+react@11.11.1_@emotion+styled@11.11.0_@types+react@18.2.18_react-dom@18.2.0_react@18.2.0/node_modules/@mui/material/Button/index.js:3
export { default } from './Button';
^^^^^^

SyntaxError: Unexpected token 'export'
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1187:20)
    at Module._compile (node:internal/modules/cjs/loader:1231:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1321:10)
    at Module.load (node:internal/modules/cjs/loader:1125:32)
    at Function.Module._load (node:internal/modules/cjs/loader:965:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:165:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:192:25)
 ELIFECYCLE  Command failed with exit code 1.
``` 
I tried following [example](https://github.com/brillout/vps-mui/) of using [vite-plugin-ssr](https://vite-plugin-ssr.com/) with [MUI](https://mui.com/). I installed ts-mode and made the required config, It doesn't work.

Also I added MUI to [ssr.noExternal](https://vite-plugin-ssr.com/broken-npm-package#errors) for the production build in my vite.config.js: 
``` 
plugins: [
            react(),
            ssr({
                // Use the default pre-render config:
                prerender: true,
                ...(process.env.NODE_ENV === 'production'
                    ? {
                          noExternal: [
                              '@mui/material',
                              '@mui/utils',
                              '@mui/base',
                              '@mui/icons-material',
                              '@mui/x-data-grid',
                              '@emotion/react',
                              '@emotion/styled',
                              'react-bootstrap',
                              '@types/react',
                              '@types/react',
                              '@types/react',
                          ],
                      }
                    : { noExternal: [] }),
            }),
           ...
``` 

Still the same result with error. So I finally proved with ['vite-plugin-cjs-interop'](https://github.com/cyco130/vite-plugin-cjs-interop):

``` 
plugins: [
            react(),
            ssr({...}),
            cjsInterop({
                // List of CJS dependencies that require interop
                dependencies: [
                    '@mui/material/*',
                    'react-bootstrap/cjs/*',
                ],
            }),
``` 
None of the above solutions works for me. [My project repository](https://github.com/Maikpwwq/Comunidad-dezzpo). What can I do to solve these errors?

```
✓ built in 2.55s
/vercel/path0/node_modules/.pnpm/@mui+material@5.14.3_@emotion+react@11.11.1_@emotion+styled@11.11.0_@types+react@18.2.18_react-dom@18.2.0_react@18.2.0/node_modules/@mui/material/Table/index.js:3
export { default } from './Table';
^^^^^
SyntaxError: Unexpected token 'export'
at internalCompileFunction (node:internal/vm:73:18)
at wrapSafe (node:internal/modules/cjs/loader:1178:20)
at Module._compile (node:internal/modules/cjs/loader:1220:27)
at Object.Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
at Module.load (node:internal/modules/cjs/loader:1119:32)
at Function.Module._load (node:internal/modules/cjs/loader:960:12)
at ModuleWrap.&lt;anonymous&gt; (node:internal/modules/esm/translators:169:29)
at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
ELIFECYCLE  Command failed with exit code 1.
```

**Reply by brillout** [reply]
@Maikpwwq The best would be two separate *minimal* reproductions, one for MUI and one for react-bootstrap. I'll have a look at it then.

**Reply by Maikpwwq** [reply]
@brillout thank you for your quick reply. Of course, this in a repo for minimal reproductions [vite-ssr-project](https://github.com/Maikpwwq/vite-ssr-project). I import a @mui/material/Button. Works in develompment, but `pnpm run build` fails with error:

```
/media/oem/MyFiles/8_DEVELOPMENT/vite-ssr-project/node_modules/.pnpm/@mui+material@5.14.4_@emotion+react@11.11.1_@emotion+styled@11.11.0_@types+react@18.2.20_react-dom@18.2.0_react@18.2.0/node_modules/@mui/material/Button/index.js:3
export { default } from './Button';
^^^^^^

SyntaxError: Unexpected token 'export'
      ....
```

Same for `pnpm create vite-plugin-ssr` with react-ts config, watch here [vite-ssr-ts-project](https://github.com/Maikpwwq/vite-ssr-ts-project).

**Reply by brillout** [reply]
@Maikpwwq I've updated the workaround https://github.com/brillout/vps-mui to use the latest MUI version and to showcase a successful use of `@mui/material/Button`. Let me know if you still run into issues.

**Reply by Maikpwwq** [reply]
@brillout So, I clone your repo vps-mui, after install everything, I run `pnpm run dev` and get me this error:

```
> @ dev /media/oem/MyFiles/8_DEVELOPMENT/vite-ssr-ts-project
> npm run server:dev

> server:dev
> ts-node ./server/index.ts

TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /media/oem/MyFiles/8_DEVELOPMENT/vite-ssr-ts-project/server/index.ts
```
I modify script in package.json to [node --loader ts-node/esm](https://github.com/TypeStrong/ts-node/issues/1997), I use ` "server": "node --loader ts-node/esm ./server/index.ts",`, now it works in development and build in production.

I was facing other issue, because I made a bad use of "noExternals" in vite.config.ts file

```
const config: UserConfig = {
  plugins: [
    react(), 
    ssr() // wrong use noExternals here
  ],
  ssr: {
    noExternal, // right use noExternals here
  },
};
}
...
```
I am going to use updates in my main code and comment soon how is going.

**Reply by brillout** [reply]
Yes it's a known issue with ts-node: #1067. I've updated the workaround repo.

**Reply by Maikpwwq** [reply]
@brillout Thank you for updates, now works when import `esm` modules.

Please could you help my to find a solution to this `layout renderer folder` [discussion](https://github.com/brillout/vite-plugin-ssr/discussions/1081)?

**Comment by brillout** [comment]
Reproduction 1: https://github.com/brillout/vps-mui/tree/reprod-1
Reproduction 2: https://github.com/brillout/vps-mui/tree/reprod-2

**Reply by schaschko** [reply]
This is the workaround I implemented:

```
// vite.config.js
const config: UserConfig = {
  plugins: [react(), ssr({ prerender: true }), mdx()],
  ...(process.env.NODE_ENV === 'production' ? { ssr: {noExternal: ["@mui/material", "@mui/utils", "@mui/base", "@mui/icons-material"]} } : {}),
}
```

Works fine, but workaround. Let's see what the mui team thinks about the issue ;)

**Reply by brillout** [reply]
Good idea, that works as well.

FYI if you want a slightly faster build time you can keep externalizing MUI while using https://github.com/cyco130/vite-plugin-cjs-interop.

**Reply by brillout** [reply]
I actually really like your workaround, I've added a mention to it in the official workaround: https://github.com/brillout/vps-mui/commit/89dd9925276ad8dab2238761332fb5f8e5efb407.

**Comment by zbynekwinkler** [comment]
You can also try 

```javascript
import { Menu } from '@mui/icons-material';
````

It works for me in a similar setup (with 5.11.16 of `@mui/icons-material`), YMMV.

**Reply by brillout** [reply]
Doesn't seem to work:

```
"MenuIcon" is not exported by "node_modules/.pnpm/@mui+icons-material@5.11.16_@mui+material@5.13.2_@types+react@18.2.6_react@18.2.0/node_modules/@mui/icons-material/esm/index.js", imported by "pages/index/index.page.tsx".
file: /home/rom/tmp/vps-mui/pages/index/index.page.tsx:2:9
1: import { Counter } from './Counter'
2: import { MenuIcon } from '@mui/icons-material' // add this
            ^
3: console.log('MenuIcon', MenuIcon)
error during build:
RollupError: "MenuIcon" is not exported by "node_modules/.pnpm/@mui+icons-material@5.11.16_@mui+material@5.13.2_@types+react@18.2.6_react@18.2.0/node_modules/@mui/icons-material/esm/index.js", imported by "pages/index/index.page.tsx".
    at error (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:2124:30)
    at Module.error (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:13463:16)
    at Module.traceVariable (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:13884:29)
    at ModuleScope.findVariable (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:12429:39)
    at Identifier.bind (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:8127:40)
    at CallExpression.bind (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:5734:28)
    at CallExpression.bind (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:9680:15)
    at ExpressionStatement.bind (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:5738:23)
    at Program.bind (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:5734:28)
    at Module.bindReferences (file:///home/rom/tmp/vps-mui/node_modules/.pnpm/rollup@3.23.0/node_modules/rollup/dist/es/shared/node-entry.js:13459:18)
 ELIFECYCLE  Command failed with exit code 1.
```

**Reply by schaschko** [reply]
At least in dev you might want to [reconsider](https://mui.com/material-ui/guides/minimizing-bundle-size/#development-environment) this approach. Named exports should work, though, I [tested it](https://github.com/brillout/vite-plugin-ssr/discussions/901#discussion-5215353).

**Reply by brillout** [reply]
Not sure what you mean, but PR welcome to improve VPS's official workaround https://github.com/brillout/vps-mui.

**Reply by zbynekwinkler** [reply]
@brillout You are importing `MenuIcon` while `Menu` is being exported.

**Reply by brillout** [reply]
I see. There is still the issue that it slows down dev though. So I'm thinking the `ssr.noExternal` workaround is still best.

**Reply by NexaSoft-SAS** [reply]
Hi @brillout, I am having troubles again, recently I updated vite to version 5 and also vike, then icons imported with @mui/icons-material are failing with message: `Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.`. Could you guide me on this please?

**Reply by lourot** [reply]
@NexaSoft-SAS does your `onRenderHtml()` render to a stream or to a string? It vaguely rings a bell. I might be wrong but I think this may happen when `onRenderHtml()` renders to a stream but your server (e.g. express) code doesn't deal with that (not using `pageContext.httpResponse.pipe()`). Or maybe it's the other way around. See related https://github.com/vikejs/vike-react/issues/38 and https://github.com/vikejs/vike-react/pull/40

First of all, are you using the latest version of Vike?

**Reply by Maikpwwq** [reply]
Hi @AurelienLourot I am using a renderToString function in my project. 

I have created a minimal reproduction repository using commands:

`npm init vike@latest`
`npm install @mui/material @emotion/react @emotion/styled`
`npm install @mui/icons-material`

Then I modified _vite.config.ts_ file to add those material pakages to `ssr: { noExternal }` according to discussion [MUI needs to be pre-processed by Vite in production](https://github.com/vikejs/vike/discussions/901).

My vike-material-icons repo here -> https://github.com/Maikpwwq/vike-material-icons 

It fails when I import and use some icons as seen in index +Page.tsx

Following error was thrown by the onRenderHtml() hook defined at /renderer/+onRenderHtml.tsx
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.

**Reply by brillout** [reply]
@NexaSoft-SAS @Maikpwwq FYI we're working on a new docs section that guides you on how to fix such ESC/CJS issue, see https://github.com/vikejs/vike/pull/1358/files#diff-6dc8908abfb8279e7fe846e46917c50bb2490dfef4972e654e143d12c8177f4eR187-R193. Let me know if that works for you.

**Reply by lourot** [reply]
@brillout in @Maikpwwq it's not `undefined` though so I don't think your link helps, or then I also don't get it myself. I have been digging this particular issue for an hour now and I'm still digging...

**Reply by brillout** [reply]
@AurelienLourot ESM/CJS issues can also lead to an import value to wrongfully be an `object`. The React patch should give you all the insights you need.

**Reply by lourot** [reply]
A simple `console.dir()` was what I needed in this case. @Maikpwwq try this:

```diff
 import { Counter } from './Counter'
-import HouseSidingIcon from '@mui/icons-material/HouseSiding'
-import DomainDisabledIcon from '@mui/icons-material/DomainDisabled'
+import { HouseSiding, DomainDisabled } from '@mui/icons-material'
 
 export { Page }
 
 function Page() {
+  console.dir(HouseSiding)
   return (
     <>
       <h1>using Material Icons</h1>
       <ul>
         <li>Rendered to HTML.</li>
         <li>
           Interactive. <Counter />
         </li>
-        <HouseSidingIcon fontSize="medium" className="mx-2 my-1" />
-        <DomainDisabledIcon fontSize="medium" className="mx-2 my-1" />
+        <HouseSiding fontSize="medium" className="mx-2 my-1" />
+        <DomainDisabled fontSize="medium" className="mx-2 my-1" />
       </ul>
     </>
   )
 }
```

**Reply by brillout** [reply]
Just updated the PR to clarify that `but got: object` is also usually a CJS/ESM issue.

**Reply by lourot** [reply]
Thanks @brillout, I noticed for me that my first reaction when seeing I needed to paste a wall of code inside the internal implementation of React was "oh no, I'll try digging with every other possible way first". I hope I'll be braver next time ;)

**Reply by lourot** [reply]
(It's a bit like when having to use a debugger: I'll always try putting 30 traces first and hope to get away with it without having to set up the debugger.)

**Reply by brillout** [reply]
I see, good to know and that makes sense. I've updated the PR to clarify that the patch is temporary and what it does.
